SSO: Update name and email on each login

dasantonym opened the pull request:
## Scope

What's changed:

- Changed the handling of the updated user payload in the driver files under `api/auth/drivers/` for LDAP, OpenID and OAuth2. The name and email fields for users now get reset to the provider's payload on each login.

## Potential Risks / Drawbacks

- Might conflict with existing expectations for behaviour on login

## Review Notes / Questions

As already stated in the issue:


- The auth_data prop that is currently being updated is being set with the non-falsy value or null which I understand to be used to delete an existing value, correct? I have not used this for the names, but rather pass the value as is (undefined if not existing on the user info) so that the existing name remains if it is not supplied. Please correct me, if that is not the case.
- I have only made the changes for the above-mentioned files because I did not understand how this is handled in SAML. This might need to be added as well and maybe you can point me to where this is done there.
- This behaviour might be undesired for some, but not for all use cases. Would you rather have this update functionality be opt-in with an env var?

---

Fixes #20362

dasantonym committed:
Update first_name, last_name and email on recurring logins via LDAP, OAuth2 and OpenID

dasantonym committed:
Move update of role in LDAP into emitFilter() call to avoid TS error

dasantonym committed:
Use const to satisfy linter

dasantonym committed:
Sign CLA

dasantonym committed:
Reformat code

dasantonym committed:
Add env var AUTH_SYNC_USER_INFO to enable setting name and email via opt-in config

dasantonym committed:
Add AUTH_SYNC_USER_INFO to docs

dasantonym committed:
Merge remote-tracking branch 'origin/main' into sso-update-name-and-email

# Conflicts:
#	api/src/auth/drivers/oauth2.ts
#	api/src/auth/drivers/openid.ts
#	contributors.yml

dasantonym committed:
Fix OAuth "typo"

dasantonym committed:
Remove "?? null" occurrences for auth_data

dasantonym commented:
Hey @aidenfoxx, thanks for the PR! I have merged the latest from `main`.

One last question though: Is there a case where the `auth_data` would be undefined, but the name and email would be defined and then `auth_data` is set to `undefined` in the update? And would that even be a problem? I suppose unless you set the value to `null`, it does not get changed in the DB, right?

aidenfoxx commented:
It's fine if "auth_data" is undefined in the update since it will be ignored, as you say, which is expected.😄

dasantonym commented:
Ok, cool! Then, a newbie question: how do I proceed now? Not sure if I have to add a changeset to the API module since this is adding a new feature of sorts.

dasantonym committed:
Merge remote-tracking branch 'origin/main' into sso-update-name-and-email

# Conflicts:
#	api/src/auth/drivers/ldap.ts

dasantonym committed:
Reformat

dasantonym committed:
Run prettier

dasantonym committed:
Revert "Run prettier"

This reverts commit 7ff70a9c1f232024fd882181e61accfa87bea096.

dasantonym committed:
Revert "Reformat"

This reverts commit 93a4c0c5a88e85d2991f77a17d6a8e02256d4de6.

dasantonym commented:
Hey @rijkvanzanten and @aidenfoxx, just a quick question if this PR can be merged or if you want me to do anything else? I'd like to merge soon it so I don't have to keep the PR up to date. Thanks!

dasantonym commented:
No problem, thanks for getting back to me! You're right, I'll make the necessary changes and report back.

madc commented:
This seems like a very helpful PR IMGO. Is there any movement on this?

dasantonym commented:
Sorry for the standstill! I didn't finish this at the end of the year, and the new one has already started quite hectically. I hope to get this sorted out within the next two weeks.

madc commented:
Thanks for the quick update and all the invested time. Greatly appreciated!

dasantonym committed:
Merge branch 'main' into sso-update-name-and-email

# Conflicts:
#	api/src/env.ts
#	contributors.yml

dasantonym committed:
Sync user info setting now available for each provider, update docs

dasantonym committed:
Merge branch 'main' into sso-update-name-and-email

dasantonym commented:
Hi @aidenfoxx & @rijkvanzanten, I have finally found the time to make the requested changes. Hope this is fine to merge now!

kulogix commented:
For anyone looking to implement this functionality sooner, you can add the following hook (e.g., /extensions/hooks/auth-update-userinfo/index.js):

```javascript
export default async ({ filter }, { logger, database, services, getSchema }) => {
	const { UsersService } = services;
	const users = new UsersService({ schema: await getSchema(), knex: database });
	
	filter('auth.update', async (input, { identifier, provider, providerPayload }) => {
		if (providerPayload.userInfo && providerPayload.userInfo.sub) {
			logger.info(`Login for user with external identifier: ${identifier}`);
			// logger.info(`userPayload.auth_data: ${JSON.stringify(input)}`);
			// logger.info(`providerPayload: ${JSON.stringify(providerPayload)}`);
			const user = await database
				.select('id', 'first_name', 'last_name', 'email')
				.from('directus_users')
				.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
				.first();
			// logger.info(`Existing User: ${JSON.stringify(user)}`);
			
			if (
				user
				&& (providerPayload.userInfo.email && providerPayload.userInfo.given_name && providerPayload.userInfo.family_name)
				&& (
					providerPayload.userInfo.email != user.email
					|| providerPayload.userInfo.given_name != user.first_name
					|| providerPayload.userInfo.family_name != user.last_name
				)
			) {
				const userIds = await users.updateByQuery(
					{ filter: { id: user.id } }, 
					{ 
						email: providerPayload.userInfo.email,
						first_name: providerPayload.userInfo.given_name, 
						last_name: providerPayload.userInfo.family_name
					}
				);
				logger.info(`During login for user with external identifier ${identifier}, updated user ${user.id}: email from '${user.email}' to '${providerPayload.userInfo.email}', first name from '${user.first_name}' to '${providerPayload.userInfo.given_name}', last name from '${user.last_name}' to '${providerPayload.userInfo.family_name}'.`);
			} else if (
				user
				&& providerPayload.userInfo.email
				&& providerPayload.userInfo.email != user.email
			) {
				const userIds = await users.updateByQuery(
					{ filter: { id: user.id } },
					{
						email: providerPayload.userInfo.email
					}
				);
				logger.info(`During login for user with external identifier ${identifier}, updated user ${user.id}: email from '${user.email}' to '${providerPayload.userInfo.email}'.`);
			}
		}
		
		// Return the payload unchanged
		return input;
	});
};
```

aidenfoxx commented on api/src/auth/drivers/oauth2.ts:
```diff
@@ -176,9 +176,24 @@ export class OAuth2AuthDriver extends LocalAuthDriver {
 		if (userId) {
 			// Run hook so the end user has the chance to augment the
 			// user that is about to be updated
+			let emitPayload;
+
+			if (syncUserInfo) {
+				emitPayload = {
+					auth_data: userPayload.auth_data,
+					first_name: userPayload.first_name,
+					last_name: userPayload.last_name,
+					email: userPayload.email,
+				};
+			} else {
+				emitPayload = {
+					auth_data: userPayload.auth_data,
+				};
+			}
```
**nitpick:**
```suggestion
			let emitPayload = {
				auth_data: userPayload.auth_data,
			};

			if (syncUserInfo) {
				emitPayload = {
					...emitPayload,
					first_name: userPayload.first_name,
					last_name: userPayload.last_name,
					email: userPayload.email,
				};
			}
```

aidenfoxx commented on api/src/auth/drivers/openid.ts:
```diff
@@ -203,9 +203,24 @@ export class OpenIDAuthDriver extends LocalAuthDriver {
 		if (userId) {
 			// Run hook so the end user has the chance to augment the
 			// user that is about to be updated
+			let emitPayload;
+
+			if (syncUserInfo) {
+				emitPayload = {
+					auth_data: userPayload.auth_data,
+					first_name: userPayload.first_name,
+					last_name: userPayload.last_name,
+					email: userPayload.email,
+				};
+			} else {
+				emitPayload = {
+					auth_data: userPayload.auth_data,
+				};
+			}
```
**nitpick:**
```suggestion
			let emitPayload = {
				auth_data: userPayload.auth_data,
			};

			if (syncUserInfo) {
				emitPayload = {
					...emitPayload,
					first_name: userPayload.first_name,
					last_name: userPayload.last_name,
					email: userPayload.email,
				};
			}
```

dasantonym committed:
Update api/src/auth/drivers/oauth2.ts

Co-authored-by: Aiden Foxx <aiden.foxx.mail@gmail.com>

dasantonym committed:
Update api/src/auth/drivers/openid.ts

Co-authored-by: Aiden Foxx <aiden.foxx.mail@gmail.com>

dasantonym committed:
Merge branch 'main' into sso-update-name-and-email

dasantonym commented:
Hey @aidenfoxx, thanks for reviewing the changes! I updated everything, but now the tests throw:

```
  Error: src/auth/drivers/oauth2.ts(188,6): error TS2353: Object literal may only specify known properties, and 'first_name' does not exist in type '{ auth_data: string | undefined; }'.
  Error: src/auth/drivers/openid.ts(215,6): error TS2353: Object literal may only specify known properties, and 'first_name' does not exist in type '{ auth_data: string | undefined; }'.
```

Definitely not a Typescript pro here, but I vaguely remember encountering this once when I started out on this PR and thought I had fixed it. Do you have an idea what to do here?

dasantonym commented:
Ah, now I remember. This is why I used the clumsy if/else for setting the auth data. I had the exact same code you suggested in an earlier commit, but if you initially set the `emitPayload` object with only the one prop, you are not allowed to add more props later...

dasantonym committed:
Provide emitPayload type to satisfy TS rules

dasantonym committed:
Merge branch 'main' into sso-update-name-and-email

dasantonym commented:
The latter suggestion did the trick, thanks a lot!

br41nslug commented:
Due to lacking edit permissions on the PR to resolve the merge conflicts and add a changeset 😬 Moved this PR to #23595
